### PR TITLE
[videoplayer] Go to Chapter Start for "Previous Chapter" Past Grace Period

### PR DESCRIPTION
**Pull request by CrystalP**
## Description

Modify VP actions when a backwards chapter skip is requested ("previous chapter", "previous chapter or big step back" actions)

Skip requested within 5 seconds of chapter start: go to the beginning of the previous chapter (same as current behavior)
Skip requested past 5 seconds from chapter start: go to the beginning of the current chapter.

(5 seconds taken from similar code used for scenes, could be adjusted)

## Motivation and context

I find the current behavior of going to the previous chapter regardless of the playing position in the current chapter unintuitive, even though it's technically correct.

For example in the situation below, skipping to previous chapter goes to the very beginning position 00:00, even though playback is almost at the end of chapter 2.
image

In addition, I didn't find a way to go back to the start of the currently playing chapter.

I think it's a nice improvement to extend the previous chapter action, and to make it go back to the start of the current chapter after a certain time has passed in the current chapter. It avoids the need to map another button on a remote or on keyboard for a "back to chapter start" function.

## How has this been tested?

Windows, but platform is not relevant.
Files with / without chapters, skip backwards before 5 seconds of playback

## What is the effect on users?

More intuitive behavior for chapter backwards skip / Down key for large step or chapter back.

## Screenshots (if appropriate):
Initial situation:
image

Current master: "previous chapter"goes to the beginning of the file
image

With PR, "previous chapter" goes to start of the current chapter:
image
An additional "previous chapter" within 5 seconds of chapter start goes to the previous chapter (=beginning of the file here).
image

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by thexai** [comment]
I like the change and I think improves usability as users surely expect skip to previous chapter allows skip to previous chapter mark. Is not intuitive skip two chapters mark back (although is technically correct). 

But current change in PR only changes this for keyboard actions up / down and not if is used OSD playback menu. To maintain coherence same changes should be applied here:

https://github.com/xbmc/xbmc/blob/22a988aad334e31230f8b7cfd6b20fc0b821586a/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L4516

and here

https://github.com/xbmc/xbmc/blob/22a988aad334e31230f8b7cfd6b20fc0b821586a/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L4389

This allows OSD buttons has same behavior.

skip

**Comment by CrystalP** [comment]
Good catch. Reimplemented as function GetPreviousChapter() used by the 3 callers.
Takes care of keyboard, OSD and remotes.

Tried testing the ```if (pMenus->IsInMenu())``` branch of OnAction, but didn't find a menu that allows the condition ```if (pMenus->CanSeek() && GetChapterCount() > 0 && GetChapter() > 0)```
A DVD returns false in CanSeek and the BR I had on hand returns CanSeek=true, GetChapterCount=1 but GetChapter=0. I think the change of the PR still makes sense for that case though.

File: xbmc/cores/VideoPlayer/VideoPlayer.h [hunk 1]
```diff
@@ -323,6 +323,7 @@ class CVideoPlayer : public IPlayer, public CThread, public IVideoPlayer,
   int GetChapter() const override;
   void GetChapterName(std::string& strChapterName, int chapterIdx = -1) const override;
   int64_t GetChapterPos(int chapterIdx = -1) const override;
+  int GetPreviousChapter();
```
Comment (thexai) [inline xbmc/cores/VideoPlayer/VideoPlayer.h:326]: Maybe move to line 372 after `void SetUpdateStreamDetails();` as all others are override and usually overrides are enumerated fist.

**Review by thexai — APPROVED** [review]

File: xbmc/cores/VideoPlayer/VideoPlayer.h (hunk 1 above)
Comment (CrystalP) [reply xbmc/cores/VideoPlayer/VideoPlayer.h]: I wasn't sure. makes sense though, I also made it protected, we tend to make too much stuff public for no reason.

**Pull request merged**